src/herder/PendingEnvelopes.cpp

```diff
@@ -440,7 +448,7 @@ PendingEnvelopes::cleanKnownData()
     auto it2 = mKnownTxSets.begin();
     while (it2 != mKnownTxSets.end())
     {
-        if (it2->second.expired())
+        if (it2->second.first.expired())
```

Comment:
maybe we should also leave a comment that we normally expect tx set frame to outlive resolved tx set for this to work properly. From what I see, at apply time, we actually don't use the cache, and re-resolve the tx set anyway. So the only time we use the cache is during consensus

Reply:
The check would be redundant/incorrect as it would mean that someone holds a reference outside of the cache, which isn't useful for maintaining known tx sets. So I don't think it's needed. Added a comment though.

> So the only time we use the cache is during consensus

Yes, that's the only place where that seems safe. At apply time we might deal with non-resolved tx sets that don't need to be cached (e.g. during catchup).

Reply:
> The check would be redundant/incorrect as it would mean that someone holds a reference outside of the cache, which isn't useful for maintaining known tx sets. So I don't think it's needed. Added a comment though.

no, I don't think that's true; the point of keeping weak references here is because SCP is holding strong references to values that might get pushed out of the PE and we need to ensure they aren't deleted

Reply:
`mKnownTxSets` only holds the weak references. It actually seems wrong to do the cleanup based on the ptr being expired or not; the check should rather be based on the cache inclusion. Otherwise you'd just hold a weak reference to an object that is potentially held outside of PE, i.e. there is no guarantee that the next `get` call would actually refer to alive object.

Reply:
From what I see, the assumption that all data needed is always present in the cache isn't right. the reason weak references are kept here is because the SCP subsystem (in addition to PE)  holds strong references to tx sets that might be evicted from the cache but are necessary for consensus. This is how SCP ensures that data needed for consensus isn't deleted (https://github.com/stellar/stellar-core/pull/2369 updated SCP to store strong references, and reworked the PE cache be able to retrieve the tx sets that got pushed out). Writing this down, I realized that the currently proposed change to SCP probably has a problem, as it asserts that the resolved tx set must be present (but it might get pushed out if SCP isn't keeping a strong reference).

Reply:
Thanks for the clarification. As we've agreed offline, let's just get rid of caching for the time being; the lifetime headache doesn't seem to be worth some minimal performance gains.  The old approach resolved every tx set anyway; we might just do that twice per tx set now, but the cost is low - it's a bit of copies and some processing linear to tx count.